stats/statsaxis.cpp

```diff
@@ -0,0 +1,394 @@
+// SPDX-License-Identifier: GPL-2.0
+#include "statsaxis.h"
+#include "statstranslations.h"
+#include "statsvariables.h"
+#include "core/pref.h"
+#include "core/subsurface-time.h"
+#include <math.h> // for lrint
+#include <numeric>
+#include <QChart>
+#include <QFontMetrics>
+#include <QLocale>
+
+StatsAxis::StatsAxis(bool horizontal) : horizontal(horizontal)
+{
+}
+
+StatsAxis::~StatsAxis()
+{
+}
+
+std::pair<double, double> StatsAxis::minMax() const
+{
+	return { 0.0, 1.0 };
+}
+
+// Guess the number of tick marks based on example strings.
+// We will use minimum and maximum values, which are not necessarily the
+// maximum-size strings especially, when using proportional fonts or for
+// categorical data. Therefore, try to err on the safe side by adding enough
+// margins.
+int StatsAxis::guessNumTicks(const QtCharts::QChart *chart, const QtCharts::QAbstractAxis *axis, const std::vector<QString> &strings) const
+{
+	QFont font = axis->labelsFont();
+	QFontMetrics fm(font);
+	int minSize = fm.height();
+	for (const QString &s: strings) {
+		QSize size = fm.size(Qt::TextSingleLine, s);
+		int needed = horizontal ? size.width() : size.height();
+		if (needed > minSize)
+			minSize = needed;
+	}
+
+	// Add space between labels
+	if (horizontal)
+		minSize = minSize * 3 / 2;
+	else
+		minSize *= 2;
+	QRectF chartSize = chart->plotArea();
+	double availableSpace = horizontal ? chartSize.width() : chartSize.height();
+	int numTicks = lrint(availableSpace / minSize);
+	return std::max(numTicks, 2);
+}
+
+ValueAxis::ValueAxis(double min, double max, int decimals, bool horizontal) : StatsAxisTemplate(horizontal),
+	min(min), max(max), decimals(decimals)
+{
+}
+
+std::pair<double, double> ValueAxis::minMax() const
+{
+	return { QValueAxis::min(), QValueAxis::max() };
+}
+
+static QString makeFormatString(int decimals)
+{
+	return QStringLiteral("%.%1f").arg(decimals < 0 ? 0 : decimals);
+}
+
+void ValueAxis::updateLabels(const QtCharts::QChart *chart)
+{
+	using QtCharts::QValueAxis;
+
+	// Avoid degenerate cases
+	if (max - min < 0.0001) {
+		max = 0.5;
+		min = -0.5;
+	}
+
+	QLocale loc;
+	QString minString = loc.toString(min, 'f', decimals);
+	QString maxString = loc.toString(max, 'f', decimals);
+	int numTicks = guessNumTicks(chart, this, { minString, maxString});
+
+	// Use full decimal increments
+	double height = max - min;
+	double inc = height / numTicks;
+	double digits = floor(log10(inc));
+	int digits_int = lrint(digits);
+	double digits_factor = pow(10.0, digits);
+	int inc_int = std::max((int)ceil(inc / digits_factor), 1);
+	// Do "nice" increments: 1, 2, 4, 5.
```

Comment:
that's not what the code does...

Reply:
Improved comment and code (yes, it is a bit indirect, but should now do what it says).